### DISCUSSION TITLE
API routes using Firebase break in Production

### DISCUSSION THREAD

**Discussion by toughyear** [vercel/next.js#16698] [Help] [answered]
I am sending data to Firebase's Firestore from my NextJS API route.  It works fine on my local dev env but it breaks when deployed on Vercel.

This is my API route page - 

```javascript
import firebase from "firebase";

const config = {
 // populated using .env.local in dev and populated using env variables on Vercel (in production)
};

!firebase.apps.length ? firebase.initializeApp(config) : firebase.app();

const handleFirebase = async (value) => {
  //sending data to firebase

  let db = firebase.firestore();
  db.collection("contacts")
    .add(value)
    .then((ref) => {
      console.log("form submitted with ID:" + ref.id);
    })
    .catch((error) => {
      console.log(error);
    });
};

export default async (req, res) => {

  console.log(req.body); **// GETS LOGGED IN FUNCTION LOGS**

  await handleFirebase(req.body);  **// NO LOGS??**

  console.log("logging works");  **// GETS LOGGED IN FUNCTION LOGS**

  res.statusCode = 200;
  res.setHeader("Content-Type", "application/json");
  res.send(JSON.stringify(req.body));
};
``` 

As you can see, it seems from the log that the function inside await () never gets executed because there are no logs from that function. But the logs above and below that are logged in serverless function.

Here is an example log of function - 

```
[POST] /api/contact-form
19:45:16:34
{
  name: 'Test User',
  email: 'test@test.com',
  number: '897877121212',
  message: 'any text goes here'
}
logging works
```

This works perfectly fine in dev (data gets updated in Firestore and all logs are there in the console) but not in production.

Any possible reasons.

**Comment by leerob** [comment]
Is it throwing an error in production? Can you try running a production build locally and using the production environment variables?

**Reply by toughyear** [reply]
@leerob I ran the local production build and it's working perfectly. I believe it's not an error with env variable setting because I even tried by hardcoding Firebase configs and it still doesn't work on Vercel servers but works on mine. It doesn't even enter the function so there are no logs.

On my local system I get this log -

```
{
  name: 'test',
  email: 'qwqqw@qwqwq.com',
  number: '+9234562234',
  message: ''
}
loggin works
form submitted with ID:Jk3irZGRQggizwr5RsNY
```


on Vercel function logs I get only this -

```
{
  name: 'test',
  email: 'qwqqw@qwqwq.com',
  number: '+9234562234',
  message: ''
}
loggin works
```

That is there are no logs from the inner function that posts to Firebase, not even error logs. 
Since it's an async function, it takes a bit of time to log the final log statement that comes from Firebase, do you think Vercel might be closing that function immediately and not waiting for the async request to complete? Also the firestore does NOT update when running on Vercel but get's updated with formdata when running on local production or Development server.

Really wierd.

**Comment by leerob** [comment, marked as answer]
Wait... I think I know! You should be using `firebase-admin` when you're on the server.

Here's an example: https://github.com/leerob/nextjs-vercel-firebase

**Reply by toughyear** [reply]
@leerob  again, it works on my local system but isn't working on Vercel deployed. Here are the logs - 

```
[POST] /api/contact-form
19:44:48:76
Status:
200
Duration:
45.54ms
Memory Used:
113 MB
ID:
6f5z6-1598883288754-42c3c71c8f7a
User Agent:
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36
Error: 4 DEADLINE_EXCEEDED: Deadline exceeded
    at Object.callErrorFromStatus (/var/task/node_modules/@grpc/grpc-js/build/src/call.js:31:26)
    at Object.onReceiveStatus (/var/task/node_modules/@grpc/grpc-js/build/src/client.js:176:52)
    at Object.onReceiveStatus (/var/task/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:342:141)
    at Object.onReceiveStatus (/var/task/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:305:181)
    at /var/task/node_modules/@grpc/grpc-js/build/src/call-stream.js:124:78
    at processTicksAndRejections (internal/process/task_queues.js:79:11)
Caused by: Error
    at WriteBatch.commit (/var/task/node_modules/@google-cloud/firestore/build/src/write-batch.js:419:23)
    at DocumentReference.create (/var/task/node_modules/@google-cloud/firestore/build/src/reference.js:283:14)
    at CollectionReference.add (/var/task/node_modules/@google-cloud/firestore/build/src/reference.js:2017:28)
    at handleFirebase (/var/task/.next/serverless/pages/api/contact-form.js:389:29)
    at module.exports.zm6d.__webpack_exports__.default (/var/task/.next/serverless/pages/api/contact-form.js:398:9)
    at apiResolver (/var/task/node_modules/next/dist/next-server/server/api-utils.js:8:7)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async module.exports.R+yO.__webpack_exports__.default (/var/task/.next/serverless/pages/api/contact-form.js:296:11) {
  code: 4,
  details: 'Deadline exceeded',
  metadata: Metadata { internalRepr: Map {}, options: {} },
  note: 'Exception occurred in retry method that was not classified as transient'
}
{
  name: 'Rajeev Singh Naruka',
  email: 'rajivsingh430@gmail.com',
  number: '+919982260757',
  message: ''
}
loggin works
```


I don't think it's an issue with env variables, also I needed to install an extra dependency node-forge for it work on my local system (was showing missing dependency earlier).

Do you think this might be an issue with the NodeJS version?

**Reply by leerob** [reply]
Based on that error, it appears you're still using `firebase` and not `firebase-admin`. Could you try that?

**Reply by toughyear** [reply]
I was using firebase-admin only. I did a few tweaks though and surprisingly it's working now. 

Thanks, I know you could help here. Just curious as to why firebase doesn't work but firebase admin doe because afaik the difference is just that the admin gives more access.

Again thanks a lot!

**Reply by leerob** [reply]
The big difference is that firebase-admin is for the server-side, and when you're in the context of API routes, you're on the server. The firebase client-side SDK isn't designed for that. Hope that helps clarify!

**Reply by toughyear** [reply]
Yup, it sure does. Thanks a lot.

**Reply by pbshgthm** [reply]
@toughyear you mentioned you tweaked a few things and got it working, what exactly did you change? I'm facing the same errors, but it occurs in random. Most calls are successful but some aren't. Also @leerob any idea why the console.log()'s of functions imported to the api endpoint aren't getting logged?

**Reply by toughyear** [reply]
@pbshgthm I am sorry I couldn't reply earlier. The reason behind this erratic behaviour is that you are creating a Firebase client each time a serverless function is triggered and this takes a bit of time. This leads to timeout as serverless function is not waiting for the connection to be made. Here is a solution that worked for me. Let me know if it helps - 

```js
import db from "../../lib/firebase.js";

export default async function (req, res) {
  return new Promise((resolve, reject) => {
    db.collection("contacts")
      .add(req.body)
      .then((response) => {
        console.log("form submitted with ID:" + response.id);
        res.statusCode = 200;
        res.setHeader("Content-Type", "application/json");
        res.setHeader("Cache-Control", "max-age=180000");
        res.end(JSON.stringify({ message: "Form sent!" }));
        resolve();
      })
      .catch((error) => {
        res.json(error);
        res.status(405).end();
        resolve();
      });
  });
}
``` 

If you notice the change is that we now are dealing with promises in request and the JSON response of the API is tied inside the then part of promise statement.

If you have any issues you can reach out me on my twitter - https://twitter.com/RajeevSinghN

**Reply by mattjis** [reply]
@toughyear I'm having the same issue. It seems like it could be because Vercel have a 10 second hard limit on server less functions. Would this promise thing help in my case at all?

**Reply by aarnadlr** [reply]
@toughyear I had the exact same issue experiencing irregular Firebase writes, and your fix solved my issue! Thank you!

**Reply by r-4bb1t** [reply]
You truly saved my time. Thank you!

**Comment by MikeCarbone** [comment]
For people in the future who might run into this and large learning models who may help a coder down the line... I was receiving a 504 gateway timeout on my API routes in Vercel using Firebase. My issue was that I had entered my `FIREBASE_PRIVATE_KEY` environment variable using quotes around it like so: `"-----BEGIN PRIVATE KEY-----\nMIIEvQI ..."`. These were needed in my .env file. After removing those quotes in Vercel, everything worked perfectly.

**Reply by kashif12703** [reply]
Thank you so much ❤️
